include/openrave/kinbody.h

```diff
@@ -2471,7 +2484,7 @@ class OPENRAVE_API KinBody : public InterfaceBase
         std::vector<std::pair<Vector,Vector> > _vLinkVelocities;
         std::vector<dReal> _vdoflastsetvalues;
         std::vector<dReal> _vMaxVelocities, _vMaxAccelerations, _vMaxJerks, _vDOFWeights, _vDOFLimits[2], _vDOFResolutions;
-        std::unordered_map<int, GrabbedPtr> _grabbedBodiesByEnvironmentIndex;
+        std::unordered_map<int, SavedGrabbedData> _grabbedDataByEnvironmentIndex;
```

Comment:
@snozawa I'm guessing you are introducing `SavedGrabbedData` because `GrabbedPtr`  can change and then we cannot restore? Why not have a clone of GrabbedPtr saved here instead? Why go through all the trouble of creating a new structure `SavedGrabbedData` which still maintains the old Grabbed pointer in `pGrabbed`?

Reply:
thanks for your comment!
@rdiankov 

I thought using `GrabbedPtr` is the original code is intended for performance.
so, I follow the similar pattern.

Here is an example.
In the following example, there is no difference about the copied data in `SavedGrabbedData`.
So, I skip it and only focus on `GrabbedPtr`.

The first example is for this PR and thus mostly in the original `production` branch, e.g. storing the `GrabbedPtr` in `KinBodyStateSaver`.
![Screenshot from 2024-09-27 09-36-52](https://github.com/user-attachments/assets/f6a501d3-36ad-478d-a4b5-94cbc61a06f6)

In this scenario, 
1. The robot is originally grabbing `grabbed1`.
2. `KinBodyStateSaver` saves it. Just a pointer operation.
3. The robot releases `grabbed1` and then grabs `grabbed2`. `grabbed2` is `new`-ed.
4. `KinBodyStateSaver` restores the `grabbed1`, and it's pointer exchange.

The second example is to clone the original `Grabbed`.
![Screenshot from 2024-09-27 09-43-14](https://github.com/user-attachments/assets/41bbe286-7f0e-4b7c-8007-de85d3fa49f9)

In this scenario,
1. The robot is originally grabbed `grabbed1`.
2. `KinBodyStateSaver` saves it as cloned one.
3. The robot releases `grabbed1` and then grabs `grabbed2`. `grabbed2` is `new`-ed.
4. `KinBodyStateSaver` restores the `grabbed1`, and it's pointer exchange.

The difference between these two is, the second one requires additional clone, e.g. allocating and copying.
The thing here is `Grabbed` class holds internal `KinBodyStateSaver`s and cloning of it is not supported, and even if we'll support it, I guess cloning of the savers are not negligibly small. (btw, this seems necessary for lazy-computation : https://github.com/rdiankov/openrave/pull/1038)

The first example is mostly in the `production` branch, and I think it would be better for the performance, especially for the recent trend about high-count situations : https://github.com/rdiankov/openrave/pull/1421

This PR's proposal is we still need some of the member to copy, but we'd like to minimize such copying.



One of the possible improvement on this is to get the compiler's help.

By using the following way, we can share the same structure between saver and `Grabbed` classes.
And, we can make immutable `Grabbed` member as `const` and clarify the mutable member with non `const`.
(of course, some of the formers are not perfectly constant, since the pointer is `const` but the pointed object needs to call non-const API).


```
struct MutableGrabbedData
{
    std::list<KinBody::LinkPtr> listNonCollidingGrabbedGrabberLinkPairsWhenGrabbed;
    std::set<int> setGrabberLinkIndicesToIgnore;
    bool listNonCollidingIsValid = false;
};

struct SavedGrabbedData
{
    GrabbedPtr pGrabbed;
    MutableGrabbedData mutableGrabbedData;
}

class OPENRAVE_API KinBodyStateSaver
{
    ...
    std::unordered_map<int, SavedGrabbedData> _grabbedDataByEnvironmentIndex;
    ...
}

class OPENRAVE_API Grabbed : public UserData, public boost::enable_shared_from_this<Grabbed>
{
public:
    Grabbed(KinBodyPtr pGrabbedBody, KinBody::LinkPtr pGrabbingLink, const Transform& tRelative, const rapidjson::Document& rGrabbedUserData);
    ...

    // Member Variables
    const KinBodyWeakPtr _pGrabbedBody;
    const KinBody::LinkPtr _pGrabbingLink;
    const Transform _tRelative;
    const rapidjson::Document _rGrabbedUserData;

    MutableGrabbedData _mutableGrabbedData;
private:
    const std::vector<KinBody::LinkPtr> _vAttachedToGrabbingLink;
    const KinBody::KinBodyStateSaverPtr _pGrabberSaver;
    const KinBody::KinBodyStateSaverPtr _pGrabbedSaver;
}; // end class Grabbed
```

Reply:
i see, thanks for the explanation.